### DISCUSSION TITLE
Unable to @entangle property that implements Wireable interface

### DISCUSSION THREAD

**Discussion by harrisbp** [livewire/livewire#4060] [7. Bugs (archived)] [answered]
- Laravel v8.65
- Livewire v2.7
- Alpinejs@3.4.2

### Steps to reproduce:

1. Base install for all of the components.
2. Create a Counter component using the make:livewire command with a property
3. Create an Options property that implements Livewire\Wireable
4. In the counter.blade.php file use @entangle to bind the component to the public $options property.
5. Click the button and the livewire error is thrown: Cannot assign array to property App\Http\Livewire\Counter::$options of type App\Options

```php
<?php
namespace App\Http\Livewire;

use Livewire\Component;
use App\Options;

class Counter extends Component
{
    public Options $options;

    public function mount()
    {
        $this->options = new Options(['count' => 1]);
    }

    public function render()
    {
        return view('livewire.counter');
    }
}
```

```php
<?php

namespace App;

use Illuminate\Support\Fluent;
use Livewire\Wireable;

class Options implements Wireable
{
    protected $obj;
    public function __construct(array $config)
    {
        $this->obj = new Fluent($config);
    }

    public function toLivewire()
    {
        return $this->obj->toArray();
    }

    public static function fromLivewire($value)
    {
        return new static($value);
    }
}
```

```html
<div x-data="{ options: @entangle('options') }">
    <h1 x-text="options.count"></h1>
    <button @click="options.count = 10">Set value to 10</button>
</div>
```

**Comment by joshhanley** [comment, marked as answer]
@harrisbp currently this is not support, but I have a work around posted on #3974 which I will copy here. You will need to change `updatingE` to `updatingOptions` and update the wireable class in the method.

Hope this helps!

---

currently the way the wireable DTO's are setup, is so that you can bind to properties on the DTO, not the whole DTO itself.

We're currently working on a solution for that, but for now, you can work around it by using an updating hook in your component and using value by reference &$value.

```php
public function updatingE(&$value) // Notice the pass by reference here
{
    $value = EstateTypeEnum::fromLivewire($value);
}
```

**Reply by harrisbp** [reply]
@joshhanley  Thank you so much for your help!

I started to go down the road to submitting a PR to fix this issue and identified the issue inside of Livewire\ComponentConcerns\HandlesActions where the value for a wireable is set on line 52 via $this->{$name} = $value;

Was I headed in the right direction?

**Reply by joshhanley** [reply]
@harrisbp yeah that's right spot 🙂

We're looking at a completely different way to manage property types and updates, etc. now though. If you are interested have a look at #3812

**Reply by adriaanzon** [reply]
@joshhanley Thanks for the tip on modifying the `update...()` methods, that helped me to get Wireables to work for me. Is there any way to make Wireable properties work with the query string? Or is this going to be tackled in #3812?

**Reply by joshhanley** [reply]
@adriaanzon no worries! Hmm not sure, what are you hoping to achieve?

**Reply by adriaanzon** [reply]
I was hoping to use Wireable properties in the query string like the example below, but the query string feature isn't currently aware of Wireable properties. Instead of using the value from `toLivewire()`, it converts the object to an array and I get this as query string: `?foo[value]=bar`

```php
class Component
{
    public Foo $foo;

    public function queryString()
    {
        return ['foo'];
    }
}

class Foo implements Wireable
{
    public string $value;

    public function toLivewire()
    {
        return $this->value;
    }

    public static function fromLivewire($value)
    {
        return new static($value);
    }
}
```

**Reply by joshhanley** [reply]
@adriaanzon a right yeah definitely won't work at the moment. Can you use another property that is bound to the querystring to track the value of your wireable and then generate it from that?

**Reply by adriaanzon** [reply]
@joshhanley Yeah, that's what I was doing, using computed properties. However, my use case is a data filter component containing multiple filters that apply an Eloquent query. As I'm adding more filters, adding all this boilerplate code gets a bit old, so I was looking for a way to refactor this using a Wireable for every filter.

But yeah, for now this is solution is okay. I have moved all code related to the individual filters into a trait, which already cleans up the component itself.

**Reply by Niq1982** [reply]
@joshhanley Any ideas how to implement this workaround so one would not have to do it in every component? Tried to do it in a Trait but it doesnt work. It sets the value on `$this`, but i still get the `HandlesActions` -error mentioned above.

```
trait WithDatepickerInput
{
    public function updatingWithDatepickerInput($key, &$value)
    {
        $class = get_class($this->$key);
        if ($class !== 'Datepicker\DatepickerInput' || !property_exists($this, $key)) {
            return;
        }
        $value = $this->$key::fromLivewire($value);
    }
}
``` 

If I rename that same function to `updating`, it works just fine. But it's not possible have two methods with the same name, so if i want to add a `updating` method to my Component, it will overwrite and ignore the `updating` method in my Trait.

**Reply by joshhanley** [reply]
@Niq1982 I don't think that would work like that in a trait. The reason is that trait methods are fired by an internal listener. Due to it passing through an event and then listener, you can't use `$value` by reference as it's not the same value object, so updating it won't actually update the original value like it does when you using a `updating` hook within a component directly.

Hope this helps!

**Reply by Perer876** [reply]
@Niq1982 There's currently a way to do it by renaming the method in the Trait so it won´t lose the reference.

In the trait:
````php
trait WithDatepickerInput
{
    public function updating($key, &$value)
    {
        $class = get_class($this->$key);
        if ($class !== 'Datepicker\DatepickerInput' || !property_exists($this, $key)) {
            return;
        }
        $value = $this->$key::fromLivewire($value);
    }
}
````
In your component:
````php
class MyComponent extends Component
{
    use WithDatepickerInput {
         updating as protected updatingWithDatepickerInput
    }

    public function updating($name, &$value): void
    {
        // Whatever you need to add.
        
        $this->updatingWithDatepickerInput($name, $value);
    }

    ...
}
````
